libraries/clchain/include/clchain/graphql_connection.hpp

```diff
@@ -0,0 +1,160 @@
+#pragma once
+
+#include <clchain/graphql.hpp>
+#include <eosio/from_bin.hpp>
+#include <eosio/reflection2.hpp>
+#include <eosio/to_bin.hpp>
+
+namespace clchain
+{
+   struct PageInfo
+   {
+      bool hasPreviousPage = false;
+      bool hasNextPage = false;
+      std::string startCursor;
+      std::string endCursor;
+   };
+   EOSIO_REFLECT2(PageInfo, hasPreviousPage, hasNextPage, startCursor, endCursor)
+
+   template <typename T, const char* ConnectionName, const char* EdgeName>
+   struct ConnectionConfig
+   {
+      using value_type = T;
+      static constexpr const char* connection_name = ConnectionName;
+      static constexpr const char* edge_name = EdgeName;
+   };
+
+   template <typename Config>
+   struct Edge
+   {
+      using config = Config;
+
+      typename Config::value_type node;
+      std::string cursor;
+   };
+   template <typename Config>
+   [[maybe_unused]] inline const char* get_type_name(Edge<Config>*)
+   {
+      return Config::edge_name;
+   }
+   template <typename Config, typename F>
+   constexpr void eosio_for_each_field(Edge<Config>*, F f)
+   {
+      EOSIO_REFLECT2_FOR_EACH_FIELD(Edge<Config>, node, cursor)
+   }
+
+   template <typename Config>
+   struct Connection
+   {
+      using config = Config;
+
+      std::vector<Edge<Config>> edges;
+      PageInfo pageInfo;
+   };
+   template <typename Config>
+   [[maybe_unused]] inline const char* get_type_name(Connection<Config>*)
+   {
+      return Config::connection_name;
+   }
+   template <typename Config, typename F>
+   constexpr void eosio_for_each_field(Connection<Config>*, F f)
+   {
+      EOSIO_REFLECT2_FOR_EACH_FIELD(Connection<Config>, edges, pageInfo)
+   }
+
+   template <typename Connection,
+             typename Key,
```

Comment:
It's probably worth noting that for the cursor to work correctly, the container cannot hold multiple equivalent keys.